### DISCUSSION TITLE
Working example of parent/child workflows using fork

### DISCUSSION THREAD

**Discussion by johnwc** [elsa-workflows/elsa-core#3281] [Q&A]
We are having issue with child workflow's status not being set as finished, staying at Running status;, causing parent workflow to stay suspended forever. This happens randomly among the child workflows, so it doesn't seem to be an issue with the child workflows faulting or such. They just sometimes never come out of Running state. I'm not sure if there is a special activity that needs to be in the child workflows to force a signal to be 'sent' to cause the workflow to be updated as Finished.

Can someone create a example parent workflow that calls 5 separate child workflows using fork activity, waits for all to complete, then using Join the activity, joins to run a few ending activities in the parent?

Parent Workflow
(Starting Activities) -> Fork Activity -> (Child Workflows run, waiting to complete) -> Join Activity -> (Finishing Activities)

Child Workflow
Can just have simple random timer activity to simulate something being done.

**Comment by johnwc** [comment]
@sfmskywalker any help here? This is a hit-or-miss whether the join activity gets activated to continue running the rest of the parent workflow. I have been restarting the workflow for a few hours now, and can't get it to run all the way.

**Reply by sfmskywalker** [reply]
I wish I could but unfortunately I’m completely swamped and tomorrow I’ll be on the road until next Wednesday. Realistically the earliest I can try and look into this is the week thereafter, but I’ll try my best to do before. In the meantime, if you can setup the workflow you described that reproduces the issue in a sample project, then I (or anyone else reading this that wants to help) can get a jump start on this.

**Reply by johnwc** [reply]
Well, I too will be out until next week. I would really like to do a screen share with you to let you visually see the issue. It would be difficult to create a test workflow, as all of our workflows utilize custom activities. But, if I get a free moment I will try to create a workflow that mimics them with pausing timers.

**Reply by johnwc** [reply]
@sfmskywalker here is an small example. The execution order is from bottom up. The first WF(v5) is nothing more than a cron activity with a start workflow after it, to schedule the workflow. The main WF(v21) splits out to 6 child workflows that each start with a 10 sec timer activity, to initiate the main workflow to suspend. As you can see in the screenshot, one of the child workflows says it is still running. But, as you can see, the main WF is no longer suspended and is running. So that child WF that says it's running, is not actually running. This has been verified as each child workflow has a final activity that sends a summary email that it is complete. More times than not, the main workflow never unsuspends and just is hung in suspended status. There's nothing I have found to do to kick the tires per se of the main WF to get it to start back. I have to cancel/delete the workflows and start them again to see if it will run all the way through.

<img width="956" alt="image" src="https://user-images.githubusercontent.com/2798441/191396423-7854ac9a-be83-45db-a59c-6bf5af481e1a.png">

**Reply by johnwc** [reply]
@sfmskywalker it happened again so I have it to show you now. All the child workflows finished, but the parent did not resume.

<img width="1006" alt="image" src="https://user-images.githubusercontent.com/2798441/191410791-fdf2a570-5ede-4794-8f25-4fb39ff7c8bb.png">

**Reply by sfmskywalker** [reply]
Thanks for that. I'll try to repro and see what gives.

**Reply by sfmskywalker** [reply]
@johnwc Question about the child workflows - you mentioned that they **start** with a timer activity - is that right? If so, then these workflows will automatically start out of their own volition **as well as** when the parent workflow schedules them (I think). Since the parent needs to initiate the child workflows, it seems to me it would be better for the child workflows to start with a non-trigger activity (e.g. WriteLine or something else) to prevent them from automatically executing.

Unless that is already the case (and you're not really starting the child workflows with a Timer)?

**Reply by johnwc** [reply]
Timer, not cron, as in a pause for x time. You mentioned before that a child workflow that is set in parent to block until complete, needed a event like this to cause the parent to "suspend".

**Reply by sfmskywalker** [reply]
Sure. But are the child workflows starting with a Timer activity? If so, then you’ll see the child workflows start automatically.

**Reply by johnwc** [reply]
I'm not really sure what you're asking. The timer simply pauses the workflow for x seconds and then resumes. It does not run the workflow every x seconds. The activity really should be renamed to pause.

**Reply by sfmskywalker** [reply]
Sorry if I’m not being clear. What I’m asking is, what is the very first activity in your child workflows? Is it the Timer activity, or something else? If not sure, a screenshot will help too. The reason I’m asking about this is because it might clarify what’s happening (or it may not). 

In the meantime of our conversation I created one parent workflow and 5 child workflows as per your instructions in the first message of this thread, and I was unable to reproduce the issue you are observing. Perhaps instead of me trying to guess what your workflows look like in order to reproduce the behavior you’re seeing, let me ask for clear steps to reproduce the issue you are seeing. This could be in the form of a simplified set of workflows, or your workflows as-is (unless they contain custom activities).

**Reply by sfmskywalker** [reply]
To be clear (hopefully ;)) the child workflows I tested with do not start with a timer, but with a WriteLine activity followed with a Timer activity to cause the workflow to suspend. Otherwise, the child workflow will execute automatically, which isn’t what you want since the child workflow needs to be executed by the parent workflow.

**Reply by johnwc** [reply]
You are seeing the issue, it is very random if the parent starts back up from being suspended.

All of the workflows start with a SetVariable activity that reads from the input passed from the parent.

<img width="215" alt="image" src="https://user-images.githubusercontent.com/2798441/193363392-2a79c55b-7df9-4f1c-ae46-838ca489f8d2.png">

**Reply by sfmskywalker** [reply]
Good, that clarifies things. In that case my workflows mimic that of yours. I’ll run more tests, but so far I tried 4 times with no issue. Tomorrow I’ll try many more times and see what happens. 

By the way, did I already ask what version you’re running? 2.8.2 or 2.9 preview? If not the preview, I recommend you try that one too, just to make sure you’re not experiencing a bug that has been fixed already.

**Reply by johnwc** [reply]
Could it be an issue with the Redis or RabbitMQ functionality. We're using MySql as the persistence, if that changes anything.

**Reply by johnwc** [reply]
We are using latest stable, so no preview.

**Reply by johnwc** [reply]
@sfmskywalker we updated to 2.10 preview, still same issue. All child workflows show as `Finished`, but parent workflow still shows as `Suspended`

**Reply by sfmskywalker** [reply]
I haven’t been able to reproduce the issue. Is there any chance you can provide a workflow that I can try out?

**Reply by johnwc** [reply]
Unfortunately, all of our workflows contain custom private activities that would be impossible to run elsewhere. If I can geta break long enough, maybe I can copy them and gut them out with thread.sleep as the execution and see if it will reproduce that way. I can then share it with you.
